packages/apollo-engine-reporting/src/schemaReporter.ts

```diff
@@ -0,0 +1,165 @@
+
+    schemaReporter
+      .reportServerInfo(sendNextWithExecutableSchema)
+      .then(({ inSeconds, withExecutableSchema }) => {
+        sendNextWithExecutableSchema = withExecutableSchema;
+        setTimeout(inner, inSeconds * 1000);
+      })
+      .catch((error: any) => {
+        // In the case of an error we want to continue looping
+        // We can add hardcoded backoff in the future,
+        // or on repeated failures stop responding reporting.
+        logger.warn(`Error in reportingServerInfo: ${error}`);
+        sendNextWithExecutableSchema = false;
+        setTimeout(inner, fallbackReportingDelayInMs);
+      });
+  }
+
+  inner();
+}
+
+interface ReportServerInfoReturnVal {
+  inSeconds: number;
+  withExecutableSchema: boolean;
+}
+
+// This class is meant to be a thin shim around the gql mutations.
+export class SchemaReporter {
+  // These mirror the gql variables
+  private readonly serverInfo: EdgeServerInfo;
+  private readonly executableSchemaDocument: any;
+  private readonly url: string;
+
+  private isStopped: boolean;
+  private readonly headers: Headers;
+
+  constructor(
+    serverInfo: EdgeServerInfo,
+    schemaSdl: string,
+    apiKey: string,
+    schemaReportingEndpoint: string | undefined,
+  ) {
+
+    this.headers = new Headers();
+    this.headers.set('Content-Type', 'application/json');
+    this.headers.set('x-api-key', apiKey);
+    this.headers.set('apollographql-client-name', 'apollo-server');
+    this.headers.set('apollographql-client-version', require('../package.json').version);
+
+    this.url =
+      schemaReportingEndpoint ||
+      'https://engine-graphql.apollographql.com/api/graphql';
+
+    this.serverInfo = serverInfo;
+    this.executableSchemaDocument = schemaSdl;
+    this.isStopped = false;
+  }
+
+  public stopped(): Boolean {
+    return this.isStopped;
+  }
+
+  public stop() {
+    this.isStopped = true;
+  }
+
+  public async reportServerInfo(
+    withExecutableSchema: boolean,
+  ): Promise<ReportServerInfoReturnVal> {
+    const { data, errors } = await this.graphManagerQuery({
+      info: this.serverInfo,
+      executableSchema: withExecutableSchema
+        ? this.executableSchemaDocument
+        : null,
+    } as ReportServerInfoVariables);
+
+    if (errors) {
+      throw new ReportingError(
+        (errors || []).map((x: any) => x.message).join('\n'),
+      );
+    }
+
+    if (!data || !data.me || !data.me.__typename) {
+      throw new ReportingError(`
+Heartbeat response error. Received incomplete data from Apollo Graph Manager.
+If this continues please reach out at support@apollographql.com.
+Got response: "${JSON.stringify(data)}"
+      `);
+    }
+
+    if (data.me.__typename === 'UserMutation') {
+      this.isStopped = true;
+      throw new ReportingError(`
+      User tokens cannot be used for schema reporting. Only service tokens.
+      Please go to https://engine.apollographql.com/graph/$INSERT_YOUR_GRAPH/settings
+      to get a token
```

Comment:
Are we calling these _tokens_ or _keys_?  The environment variable is `APOLLO_KEY` and we call these _API keys_ within the UI we're referring them to.

Let's give our error messages the same care and treatment we'd give UI/product in terms of sentence fragments, punctuation, consistency, etc.  Suggestion:

```suggestion
      This server was configured with an API key for a user.  Only a service's API key may be used for schema reporting.
      Please visit the settings for this graph at https://engine.apollographql.com/ to obtain an API key for a service.
```

I'd also prefer we not include a link to the graph if it's not a functioning link when I click on it in my terminal, but instead provide instructions on how to get to the correct place.  I suspect the real answer here is to have a `s.apollo.dev` short-link to documentation for this feature.